docs/arcadia-users-group/20221011-developer-productivity/lesson.md

```diff
@@ -0,0 +1,172 @@
+
+```{bash}
+cd ~/Desktop
+git clone git@github.com:Arcadia-Science/arcadia-computational-training.git
+cd arcadia-computational-training
+ls
+```
+
+## IDEs
+
+### What are IDEs?
+
+People think of programming as mostly editing the source code of a web application or a script or a computational pipeline. But there are many other aspects of programming that happen all the time that we take for granted: debugging, testing, checking the code for errors, typos etc.
+
+Integrated development environments or IDEs bring all these aspects of programming together and offer convenient features like autocomplete, syntax highlighting, linting, code execution etc.
+
+IDEs are not necessary for programming, but with basic configuration, they can significantly increase coding productivity.
+
+### Which IDE should I use?
+
+There are many IDEs or IDE-like text editors out there. Some are language-specific (think [IntelliJ](https://www.jetbrains.com/idea/) for Java). Some are generic like [Sublime Text](https://www.sublimehq.com/). Up until very recently, I've exclusively used Sublime Text and still love it for light-weight programming.
+
+But at Arcadia Science, we'll be using [Visual Studio Code](https://code.visualstudio.com/) (aka VS Code). There are couple reasons for this:
+
+- It has a tight-integration with git and Github.
+- It has a rich marketplace of free extensions, enabling you to configure your IDE to your heart's content.
+- It works cross-platform and has a straightforward way of syncing your settings across devices.
+- It is quite easy to configure because it suggests improvements on its own as needed.
+- It has access to a built-in terminal. Try hitting CTRL + ` when you install it.
+- It has tooling built-in to enable remote SSH access. So you can use an EC2 instance and navigate it through these tools.
+
+### Installing VS Code
+
+Download and install VS Code by going to [this URL](https://code.visualstudio.com/download). Once installed, turn on sync across devices and login with your GitHub account.
+
+## Customizing VS Code
+
+You can just use VS Code as is. This section suggests a couple base-level customizations to make your life a little easier.
+
+### Settings
+
+**Note:** There are hundreds of settings you can customize, we'll only focus on the essentials.
+
+Open up the settings panel by hitting Command + , or by selecting "Code" -> "Preferences" -> "Settings" from the top left corner. There are two ways to adjust settings:
+
+1. You can search through them using the search bar
+2. You can edit the underlying JSON blob that specifies the user settings.
+
+Today, we'll mostly do #1, but if you're interested in following the second method, a simplified version of all my settings can be found in [this file](config-and-dotfiles/vscode_settings.json). So, let's get started.
+
+- **Auto Save:** Search for "Auto Save" in the settings menu. And adjust it to the non-default value (which is none). Mine is "afterDelay".
```

Comment:
What does this do? auto save after 5 min or something? or when you delay in typing?